### DISCUSSION TITLE
PortalDots独自の構成について

### DISCUSSION THREAD

**Discussion by kai0310** [portal-dots/PortalDots#865] [Q&A] [answered]
Laravelの標準として、`App\`もしくは、8.xでは`App\Models\` 直下にModelが配置されます。しかし、PortalDotsでは、Modelは `App\Eloquents\`直下に置かれているなど、PortalDots独自の構成が成されています。

PortalDotsの前身となるプロジェクトがLaravelでは書かれていなかったといった開発の背景があり、現在のような構成になっているかと思われるのですが、途中からコントリビュートするにはハードルが高いと思い、これらの構成を明示することに意味があるのではないかと思い、このdiscussionを作成しました。

**Comment by SofPyon** [comment]
@kai0310 
ありがとうございます！
確かに、Laravel 8.x では `App\Models` が標準になったので、標準とは異なる構成になっていることを明示した方がよさそうですね！

ちなみに、 @kai0310 さん自身 `App\Eloquents\` にモデルがある点について迷われた経緯はありますでしょうか？

**Reply by kai0310** [reply]
PortalDotsの場合、Laravel 6.x を現在は使用しているので、Modelは`App\`直下にあるであろうという認識がありました。また、artisanコマンドでModelを作成する場合は、`App\`直下に生成されるが、他のModelが見当たらないという点でまず迷いました。

EloquentsにModelが置かれていると分かっていれば、`$ php artisan make:model Eloquents/Model` で作成できるとは言え、親切ではないかなと思いました。
また、これは完全に僕の主観になってしまうのですが、`Eloquents `という名前からして、`Illuminate\Database\Eloquent` を拡張しているのかな？という疑問もありました。

**Reply by SofPyon** [reply]
@kai0310 
> EloquentsにModelが置かれていると分かっていれば、$ php artisan make:model Eloquents/Model で作成できるとは言え、親切ではないかなと思いました。

確かにそうですよね！ コマンドを打つときわざわざ `Eloquents` と入れないといけないということは分かりづらいですね。
これについては、 `php artisan make:model Model` と打つだけで `Eloquents` 内にモデルを作ってくれれば嬉しい感はありますね。

> また、これは完全に僕の主観になってしまうのですが、Eloquents という名前からして、Illuminate\Database\Eloquent を拡張しているのかな？という疑問もありました。
 
僕自身が、モデル ＝ Eloquent だと思っていたのでこの命名になっています。ここは Models という命名にしておけばよかったですね。

とはいえ、Laravel ユーザーから見れば `Eloquent` フォルダの中にモデルが入っているだろうという想像はできそうな気もします。
（Eloquent を拡張したい場合にわざわざフォルダは作らないと考えられる + Eloquent を拡張するという使い方はあまり一般的ではない？）

-----

ちなみに、Eloquents に限らず、PortalDots では他ではみられない特殊なフォルダがあります。（例えば `Exports` `GridMakers` など）
それぞれのフォルダは下記の意味を持っています。これらについても説明しておきたい感じがあります。

- `app/Exports` : CSV 書き出し機能を提供するためのコード。詳細は [Laravel Excel](https://laravel-excel.com) のドキュメントを参照。
- `app/GridMakers` : スタッフモードにおいて、データベースのデータを表形式(Grid)で表示する方法を定義するためのコード。

これらの説明について、どういう形式でまとめた方が開発者にわかりやすいかアイデアがあればぜひお聞きしたいです！

**Reply by kai0310** [reply]
@SofPyon 

既存の `/docs` が何に使っているのか分からないですが、リポジトリ内にMarkdownで置いて行くのがいいと思います！
そこまで管理したくない場合は、このdiscussionにまとめておくとかでも👀

**Reply by SofPyon** [reply]
@kai0310 
ありがとうございます！
`/docs` が良さそうです！

現状、ユーザー向けと開発者向けの説明が全部 README.md にまとめっていて読みづらい感じがあるので、今ある開発者向け説明も docs に移すことを考えてみます！

**Comment by SofPyon** [comment, marked as answer]
本件ですが、「PortalDotsマニュアル」のリニューアルを行なっており、開発者向けのドキュメントも「マニュアル」内に書くことにしました！

- マニュアルのURL: http://docs.portaldots.com

マニュアル内の「開発に貢献する」セクション内で、PortalDotsのフォルダ構造に関する説明を書こうと思っていますので、しばらくお待ちください！

`ユーザー向けと開発者向けの説明が全部 README.md にまとめっていて読みづらい` 問題についても一緒に解決できればと思っております。

<img width="332" alt="image" src="https://user-images.githubusercontent.com/6687653/161585620-3f31d62d-9709-4116-b176-f1710ba48373.png">

**Comment by SofPyon** [comment]
@kai0310 
PortalDotsのディレクトリ構造を https://docs.portaldots.com/contributing/directories に記載しました！

過不足などあれば、画面右の「GitHubで編集」リンクよりPull Requestを作成していただけると助かります！

**Reply by kai0310** [reply]
@SofPyon 
返信遅れてすみません 🙇‍♂️

確認できました！ドキュメントもOSSで貢献できるのがまた良いですね.....✨
時間がある際に進めてみようと思います 🙌
